lib/logstash/codecs/kms.rb

```diff
@@ -0,0 +1,140 @@
+
+  # If set, incoming messages that cannot be decrypted will try to pass the original message
+  # through this plugin's codec. By default, undecryptable messages will fail immediately.
+
+  # This is useful to avoid issues while you are migrating to an encrypted transport.
+  # For example, if you are currently sending unencrypted data, but want to start sending KMS encrypted data,
+  # there will be a window during your deployment where old hosts are sending decrypted data while new hosts
+  # are sending encrypted data.
+  # You can handle this case by setting fallback_if_invalid_format to true.
+  #
+  # The following example will pass messages that fail to be decrypted through the json codec.
+  # [source,ruby]
+  #     stdin {
+  #         codec => kms {
+  #             codec => "json"
+  #             fallback_if_invalid_format => true
+  #         }
+  #     }
+  #
+  config :fallback_if_invalid_format, :validate => :boolean, :default => false
+
+  def register
+    @crypto_client = com.amazonaws.encryptionsdk::AwsCrypto.new
+
+    credentials = com.amazonaws.auth::DefaultAWSCredentialsProviderChain.new
+    if @access_key and @access_key
+      credentials = com.amazonaws.auth::AWSStaticCredentialsProvider.new(
+        com.amazonaws.auth::BasicAWSCredentials.new(
+          @access_key,
+          @secret_key
+        )
+      )
+    elsif @aws_profile
+      credentials = com.amazonaws.auth.profile::ProfileCredentialsProvider.new(@aws_profile)
+    end
+
+    @key_provider = com.amazonaws.encryptionsdk.kms::KmsMasterKeyProvider.new(
+      credentials,
+      com.amazonaws.regions::RegionUtils::getRegion(@region),
+      com.amazonaws::ClientConfiguration.new,
+      @key_ids
+    )
+  end # def register
+
+  def decode(data)
+    begin
+      response = get_crypto_client().decryptData(@key_provider, data.to_java_bytes)
+      context = response.getEncryptionContext()
+      @encryption_context.each do |key, value|
+        if not context.containsKey(key) or context[key] != value
+          raise RuntimeError.new('Encryption context does not match expected. Recieved context: ' + context.to_s)
+        end
+      end
+      data = String.from_java_bytes(response.getResult())
+    rescue com.amazonaws.encryptionsdk.exception::AwsCryptoException
+      unless @fallback_if_invalid_format
+        raise
+      end
+    end
+
+    @codec.decode(data) do |message|
+      yield message
+    end
+  end # def decode
+
+  # Encode a single event, this returns the raw data to be returned as a String
+  def encode_sync(event)
+    data = @codec.multi_encode([event])[0][1]
+    encrypted = get_crypto_client().encryptData(@key_provider, data.to_java_bytes, @encryption_context).getResult();
+    return String.from_java_bytes(encrypted).force_encoding('BINARY')
+  end # def encode_sync
+
+  # Simple getter for the crypto_client, so that is can be mocked out in unit tests
+  def get_crypto_client()
+    return @crypto_client
+  end
```

Comment:
The canonical way to do this in Ruby:

```ruby
attr_reader :crypto_client  # generates a getter for @crypto_client

def my_method
  self.crypto_client.do_something
```

Moreover, looking at the [tests within the SDK](https://github.com/awslabs/aws-encryption-sdk-java/blob/master/src/test/java/com/amazonawsencryptionsdk/AwsCryptoTest.java), can we get higher fidelity by mocking the key provider instead of the crypto client?